File: public/problem-assets/splitmix64.rb
Data_ = Struct.new(:arr, :query)

def generate(n, q, seed)

    next_int = lambda do ||
    seed = (seed + 0x9e3779b97f4a7c15) & 0xffffffffffffffff
    seed = ((seed ^ (seed >> 30)) * 0xbf58476d1ce4e5b9) & 0xffffffffffffffff
    seed = ((seed ^ (seed >> 27)) * 0x94d049bb133111eb) & 0xffffffffffffffff
    seed = (seed ^ (seed >> 31)) & 0xffffffffffffffff
  end

  # Create 1-based indexed arrays
  arr = Array.new(n + 1, 0)
  query = Array.new(q + 1) { Array.new(4, 0) }

  (1..n).each do |i|
    arr[i] = next_int.call() % 1_000_000 + 1
  end

  (1..q).each do |j|
    t = next_int.call() % 4 + 1
    query[j][0] = t

    l = next_int.call() % n + 1
    r = next_int.call() % n + 1

    l, r = r, l if l > r

    query[j][1] = l
    query[j][2] = r

    m = t <= 2 ? 1_000_000 : 998_244_353
    query[j][3] = next_int.call() % m + 1
  end

  Data_.new(arr, query)
end

File: public/problem-assets/splitmix64.rs
pub struct Data {
    pub arr: Vec<usize>,
    pub query: Vec<[usize; 4]>,
}

pub fn generate(n: usize, q: usize, mut seed: u64) -> Data {
    let mut splitmix64 = || {
        seed = seed.wrapping_add(0x9e3779b97f4a7c15_u64);
        seed = (seed ^ (seed >> 30)).wrapping_mul(0xbf58476d1ce4e5b9_u64);
        seed = (seed ^ (seed >> 27)).wrapping_mul(0x94d049bb133111eb_u64);
        seed = seed ^ (seed >> 31);
        seed
    };

    let mut arr = vec![0; n + 1];
    let mut query = vec![[0; 4]; q + 1];

    for v in arr.iter_mut().skip(1) {
        *v = (splitmix64() % 1_000_000) as usize + 1;
    }

    for v in query.iter_mut().skip(1) {
        let t = (splitmix64() % 4) as usize + 1;
        v[0] = t;

        let mut l = (splitmix64() % n as u64) as usize + 1;
        let mut r = (splitmix64() % n as u64) as usize + 1;

        if l > r {
            std::mem::swap(&mut l, &mut r);
        }

        v[1] = l;
        v[2] = r;

        let m = if t <= 2 { 1_000_000 } else { 998_244_353 };
        v[3] = (splitmix64() % m as u64) as usize + 1;
    }

    Data { arr, query }
}
